docs/tutorials/multimodal/semantic_search.md

```diff
@@ -0,0 +1,370 @@
+
+### 6. Hybrid BM25
+
+We are proposing a new method of search ranking called *Hybrid BM25*, which combines BM25 and semantic embedding for scoring. The key idea is to use BM25 as the first-stage retrieval method (say it recalls 1000 documents for each query), then use a pretrained language model (PLM) to score all the recalled documents (1000 documents). 
+
+We then rerank the retrieved documents with the score calculated as:
+$$ score = \beta * normalized\_BM25 + ( 1 - \beta) * score\_of\_plm $$
+where 
+$$ normalized\_BM25(q_i, D_j) = \frac{\textsf{BM25}(q_i,D_j) - \min_{a\in \mathcal{Q},b\in\mathcal{D}}(\textsf{BM25}(a,b))}{\max_{a\in \mathcal{Q},b\in\mathcal{D}}(\textsf{BM25}(a,b)) - \min_{a\in \mathcal{Q},b\in\mathcal{D}}(\textsf{BM25}(a,b))},$$
+and $ \beta $ is a tunable parameter, which we will default to $0.3$ in our tutorial.
+
+#### 6.1 Defining functions
+
+
+```python
+def hybridBM25(docs_text, queries_text, docs_id, queries_id, recall_num, top_k, beta):
+    # Recall documents with BM25 scores
+    tokenized_corpus = tokenize_corpus(docs_text)
+    bm25_model = BM25Okapi(tokenized_corpus, k1=1.2, b=0.75)
+    bm25_scores = rank_documents_bm25(queries_text, queries_id, docs_id, recall_num, bm25_model)
+    
+    all_bm25_scores = [score for scores in bm25_scores.values() for score in scores.values()]
+    max_bm25_score = max(all_bm25_scores)
+    min_bm25_score = min(all_bm25_scores)
+
+    q_embeddings = map_id_content(queries_id, queries_embeddings)
+    d_embeddings = map_id_content(docs_id, docs_embeddings)
+    
+    results = {qid: {} for qid in queries_id}
+    for idx, qid in enumerate(queries_id):
+        rec_docs = bm25_scores[qid]
+        rec_doc_emb = [d_embeddings[doc_id] for doc_id in rec_docs.keys()]
+        rec_doc_id = [doc_id for doc_id in rec_docs.keys()]
+        rec_doc_emb = torch.stack(rec_doc_emb)
+        scores = cos_sim(q_embeddings[qid], rec_doc_emb)
+        scores_top_k_values, scores_top_k_idx = get_top_k(scores, top_k)
+
+        for doc_idx, score in zip(scores_top_k_idx[0], scores_top_k_values[0]):
+            doc_id = rec_doc_id[int(doc_idx)]
+            # Hybrid scores from BM25 and cosine similarity of embeddings
+            results[qid][doc_id] = \
+                (1 - beta) * float(score.numpy()) \
+                + beta * (bm25_scores[qid][doc_id] - min_bm25_score) / (max_bm25_score - min_bm25_score)
+    
+    return results
+
+
+def evaluate_hybridBM25(recall_num, beta, top_k):
+    results = hybridBM25(docs_text, queries_text, docs_id, queries_id, recall_num, top_k, beta)
+    ndcg = ir_metrics(qrel_dict, results)
+    
+    print("NDCG@10 for Hybrid BM25: ", ndcg["NDCG@10"])
+    return ndcg
+```
+
+
+```python
+recall_num = 1000
+beta = 0.3
+top_k = 10
+evaluate_hybridBM25(recall_num, beta, top_k)
+```
+
+We were able to improve the NDCG@10 score from 0.38057 to 0.38439 by tuning `beta` and `recall_num`.
+
```

Comment:
Adding a section to compare the top 3 or 5 recalled documents of different algorithms will make it easier to understand the benefits of semantic search and the hybrid method.